pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -18,149 +18,506 @@ import (
+	topologies map[uint64]*TopologyGroup
+	// Anti-affinity works both ways (if a zone has a pod foo with anti-affinity to a pod bar, we can't schedule bar to
+	// that zone, even though bar has no anti affinity terms on it. For this to work, we need to separately track the
+	// topologies of pods with anti-affinity terms, so we can prevent scheduling the pods they have anti-affinity to
+	// in some cases.
+	inverseTopologies map[uint64]*TopologyGroup
+	constraints       *v1alpha5.Constraints
+}
+
+func NewTopology(kubeClient client.Client, constraints *v1alpha5.Constraints) *Topology {
+	return &Topology{
+		kubeClient:        kubeClient,
+		constraints:       constraints,
+		topologies:        map[uint64]*TopologyGroup{},
+		inverseTopologies: map[uint64]*TopologyGroup{},
+	}
 }
 
-// Inject injects topology rules into pods using supported NodeSelectors
-func (t *Topology) Inject(ctx context.Context, constraints *v1alpha5.Constraints, pods []*v1.Pod) error {
-	// Group pods by equivalent topology spread constraints
-	topologyGroups := t.getTopologyGroups(pods)
-	// Compute spread
-	for _, topologyGroup := range topologyGroups {
-		if err := t.computeCurrentTopology(ctx, constraints, topologyGroup); err != nil {
-			return fmt.Errorf("computing topology, %w", err)
+type matchingTopology struct {
+	topology              *TopologyGroup
+	controlsPodScheduling bool
+	selectsPod            bool
+}
+
+// Initialize scans the pods provided and creates topology groups for any topologies that we need to track based off of
+// topology spreads, affinities, and anti-affinities specified in the pods.
+func (t *Topology) Initialize(ctx context.Context, pods ...*v1.Pod) error {
+	var errs error
+	errs = multierr.Append(errs, t.trackExistingAntiAffinities(ctx))
+	for _, p := range pods {
+		// need to first ensure that we know all of the topology constraints.  This may require looking up
+		// existing pods in the running cluster to determine zonal topology skew.
+		if err := t.trackTopologySpread(ctx, p); err != nil {
+			errs = multierr.Append(errs, fmt.Errorf("tracking topology spread, %w", err))
 		}
-		for _, pod := range topologyGroup.Pods {
-			domain := topologyGroup.NextDomain(constraints.Requirements.Add(v1alpha5.NewPodRequirements(pod).Requirements...).
-				Get(topologyGroup.Constraint.TopologyKey).
-				Values())
-			pod.Spec.NodeSelector = functional.UnionStringMaps(pod.Spec.NodeSelector, map[string]string{topologyGroup.Constraint.TopologyKey: domain})
+		if err := t.trackPodAffinityTopology(ctx, p); err != nil {
+			errs = multierr.Append(errs, fmt.Errorf("tracking affinity topology, %w", err))
+		}
+	}
+	return errs
+}
+
+// trackExistingAntiAffinities is used to identify pods with anti-affinity terms so we can track those topologies.  We
+// have to look at every pod in the cluster as there is no way to query for a pod with anti-affinity terms.
+func (t *Topology) trackExistingAntiAffinities(ctx context.Context) error {
```

Comment:
I typically try to avoid unnecessary context in method names. In this case, how does `existing` clarify the meaning? Are there nonexisting antiaffinities? Should `topologySpread` or `affinity` be prefixed with `existing` as well?
```suggestion
func (t *Topology) trackAntiAffinities(ctx context.Context) error {
```